apps/sms/views/shared/js/system_message_handler.js

```diff
@@ -0,0 +1,341 @@
+     * @returns {Promise.<{raw: string}|string>} L10n object that will serve as
+     * notification body.
+     * @private
+     */
+    getNotificationBody(message) {
+      // In case of sms, we want to display text content in notification.
+      if (message.type === 'sms') {
+        return Promise.resolve({ raw: message.body || '' });
+      }
+
+      var mmsNotificationBodyL10n = null;
+
+      // Here we can only have one sender, so deliveryInfo[0].deliveryStatus
+      // => message status from sender.
+      if (message.deliveryInfo[0].deliveryStatus !== 'success') {
+        // If the delivery status is manual/rejected/error, we need to apply
+        // specific text to notify user that message is not downloaded.
+        mmsNotificationBodyL10n = 'notDownloaded-title';
+      } else if (message.subject) {
+        // If subject exist, we display subject in notification.
+        mmsNotificationBodyL10n = { raw: message.subject };
+      }
+
+      return mmsNotificationBodyL10n ?
+        Promise.resolve(mmsNotificationBodyL10n) :
+        SMIL.parse(message).then((slideArray) => {
+          // If MMS has at least one non-empty text slide, let's display that
+          // text in notification otherwise use predefined 'mms message'.
+          var textSlide = slideArray.find((slide) => !!slide.text);
+          return textSlide ? { raw: textSlide.text } : 'mms-message';
+        });
+    },
+
+    /**
+     * Prepares notification title for the message
+     * @param {MozMmsMessage|MozSmsMessage} message Message to prepare
+     * notification title for.
+     * @returns {Promise.<{raw: string}|{id: string, args: Object}>} L10n object
+     * that will serve as notification title.
+     * @private
+     */
+    getNotificationTitle(message) {
+      return Contacts.findByAddress(message.sender).then((contacts) => {
+        var sender = contacts.length && contacts[0].name &&
+          contacts[0].name.length && contacts[0].name[0] ||
+          message.sender;
+
+        if (Settings.hasSeveralSim() && message.iccId) {
+          return Utils.getSimNameByIccId(message.iccId).then(
+            (simName) => ({
+              id: 'dsds-notification-title-with-sim',
+              args: { sim: simName, sender: sender }
+            })
+          );
+        }
+
+        return { raw: sender };
+      });
+    },
+
+    /**
+     * Retrieves correct application icon to use within notifications.
+     * @returns {string} Application icon URI.
+     * @private
+     */
+    getNotificationIcon() {
+      return navigator.mozApps.getSelf().then(
+        (app) => NotificationHelper.getIconURI(app)
+      );
+    },
+
+    /**
+     * Processes received class-0 message.
+     * @param {MozMmsMessage|MozSmsMessage} message Received class-0 message.
+     * @returns {Promise.<void>} Promise that is resolved once message is fully
+     * handled.
+     * @private
+     */
+    onMessageZeroClassReceived(message) {
```

Comment:
nit: Since we also call class 0 message as flash message, maybe we can name it like `onFlashMessageReceived`?